snapcraft/internal/repo/_deb.py

```diff
@@ -186,18 +188,74 @@ def get_packages_for_source_type(cls, source_type):
+                sys.getfilesystemencoding())
+        if arch in native_arch:
+            return
+        foreign_archs = subprocess.check_output(
+            ['dpkg', '--print-foreign-architectures']).decode(
+                sys.getfilesystemencoding())
+        if arch in foreign_archs:
+            return
+        logger.info('Adding foreign architecture {}'.format(arch))
+        subprocess.check_output(
+            ['sudo', 'dpkg', '--add-architecture', arch])
+
+    @classmethod
+    def _setup_multi_arch_sources(cls, apt_cache, package_name):
+        """If the given package has an arch suffix:
+        Generate arch-specific sources list
+        Update the package cache to pull in the new packages
+        Verify that the package is in the cache
+        """
+        name, arch, version = repo.get_pkg_name_parts(package_name)
+        if not arch or arch == ':native':
+            return False
+        arch = arch[1:]
+
+        sources = _get_local_sources_list()
+        if '[arch={}]'.format(arch) not in sources:
+            logger.info('Adding sources for {}'.format(arch))
+            release = platform.linux_distribution()[2]
+            sources = _format_sources_list(None, deb_arch=arch,
+                                           release=release, foreign=True)
+            with tempfile.NamedTemporaryFile() as sources_arch_file:
+                sources_arch_file.write(sources.encode())
+                sources_arch_file.flush()
+                sources_lists = os.path.join('/etc/apt/sources.list.d/',
+                                             'ubuntu-{}.list'.format(arch))
+                subprocess.check_call(['sudo', 'cp',
+                                       sources_arch_file.name, sources_lists])
+                subprocess.check_call(['sudo', 'chmod',
+                                       '644', sources_lists])
+
+        update_output = subprocess.check_output(
+            ['sudo', 'apt-get', 'update'],
+            stderr=subprocess.STDOUT).decode(sys.getfilesystemencoding())
+        # Failure to download doesn't return an error code
+        if 'Err:' in update_output:
+            raise subprocess.CalledProcessError(
+                100, ['sudo', 'apt-get', 'update'], update_output)
+        return package_name in apt_cache
+
+    @classmethod
+    def install_build_packages(cls, package_names, arch):
         """Install build packages on the building machine.
 
         :return: a list with the packages installed and their versions.
 
         """
-        new_packages = []
+        new_packages = set(cls._get_build_deps(package_names, arch))
         with apt.Cache() as apt_cache:
             try:
                 cls._mark_install(apt_cache, package_names)
             except errors.PackageNotFoundError as e:
-                raise errors.BuildPackageNotFoundError(e)
+                if cls._setup_multi_arch_sources(apt_cache, e.package):
+                    pass  # FIXME?
```

Comment:
The PR description doesn't indicate that further work needs to be done. The FIXME isn't particularly helpful, either. What happens if a release contains this before it's fixed? Should we raise an NotImplementedError or something instead?

Reply:
Good catch, I didn't mean to leave that there. It was supposed to be a reminder for myself as I had to re-think the code after merging, but I apparently forgot. Will fix it.